src/sdam/test.rs

```diff
@@ -0,0 +1,66 @@
+use std::time::{Duration, Instant};
+
+use bson::doc;
+use tokio::sync::RwLockWriteGuard;
+
+use crate::{
+    test::{FailCommandOptions, FailPoint, FailPointMode, TestClient, CLIENT_OPTIONS, LOCK},
+    Client,
+};
+
+#[cfg_attr(feature = "tokio-runtime", tokio::test(threaded_scheduler))]
+#[cfg_attr(feature = "async-std-runtime", async_std::test)]
+async fn min_heartbeat_frequency() {
+    let _guard: RwLockWriteGuard<_> = LOCK.run_exclusively().await;
+
+    let mut setup_client_options = CLIENT_OPTIONS.clone();
+    setup_client_options.hosts.drain(1..);
+    setup_client_options.direct_connection = Some(true);
+
+    let setup_client = TestClient::with_options(Some(setup_client_options.clone()), true).await;
+
+    if !setup_client.supports_fail_command().await {
+        println!("skipping min_heartbeat_frequency test due to server not supporting fail points");
+        return;
+    }
+
+    if setup_client.server_version_lt(4, 9) {
+        println!("skipping min_heartbeat_frequency test due to server version being less than 4.9");
+        return;
+    }
+
+    let fp_options = FailCommandOptions::builder()
+        .app_name("SDAMMinHeartbeatFrequencyTest".to_string())
```

Comment:
Interesting, I've never seen this before! I assume that this makes it possible to only fail commands for a certain client (assuming we use unique app names). Would we be able to use this to remove the need for a global test lock, or is this only supported on newer server versions?

Reply:
It's a relatively new option (introduced in 4.2.9, started working for initial handshakes in 4.9). Due to those version restrictions, I don't think we can use it to replace the locks unfortunately.